src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -255,8 +256,57 @@ private void RecordDependency(CallGraphNode callerKey, CallGraphNode calledKey,
         /// </summary>
         public List<ImmutableArray<CallGraphNode>> GetCallCycles()
         {
-            // ToDo: need to implement a robust algorithm for finding cycles
-            throw new NotImplementedException();
+            var callStack = new Dictionary<CallGraphNode, CallGraphNode>();
+            var finished = new HashSet<CallGraphNode>();
+            var cycles = new List<ImmutableArray<CallGraphNode>>();
+
+            void processDependencies(CallGraphNode node)
+            {
+                if (_Dependencies.TryGetValue(node, out var dependencies))
```

Comment:
To decrease nesting, it might be better to do something like
```
if (!_Dependencies.TryGetValue(node, out var dependencies))
{
    finished.Add(node);
    return;
}

// ...
finished.Add(node);

Reply:
The nesting isn't too bad here, and I prefer to have the logic that makes use of the out variable be in the conditional block. It makes the logic cleaner and more encapsulated.